spec/lib/school_code_generator_spec.rb

```diff
@@ -0,0 +1,23 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe SchoolCodeGenerator do
+  describe '.generate' do
+    it 'uses Random#rand to generate a random number up to the maximum' do
+      random = instance_double(Random)
+      allow(random).to receive(:rand).with(SchoolCodeGenerator::MAX_CODE).and_return(123)
+      allow(Random).to receive(:new).and_return(random)
+
+      expect(described_class.generate).to eq('00-01-23')
+    end
+
+    it 'generates a string containing 3 pairs of digits' do
+      expect(described_class.generate).to match(/\d\d-\d\d-\d\d/)
+    end
+
+    it 'generates a different code each time' do
```

Comment:
I think there's a small chance that this algorithm could generate the same code twice in a row, which makes this test potentially flakey. I wonder if `generate` could take an optional seed to pass to [Random.new](https://docs.ruby-lang.org/en/master/Random.html) so that the test could be made deterministic?

Reply:
Or you could stub `Rand.new` thinking about it - as we don't really need to test that method.

Reply:
Good shout :+1:

Reply:
Hmm, having tried a few things I'm less sure how useful it is.

We could stub `Random#rand` but then this feels like it's duplicating the first test:

```ruby
it 'generates a different code each time' do
  random = instance_double(Random)
  allow(random).to receive(:rand).and_return(123, 456)
  allow(Random).to receive(:new).and_return(random)

  expect(described_class.generate).not_to eq(described_class.generate)
end
```

Or we could stub `Random.new_seed` (which is used to set the `seed` if no explicit seed is passed to `Random.new`) but I don't think that's enough to guarantee that we won't get duplicate numbers, is it?

```ruby
it 'generates a different code each time' do
  allow(Random).to receive(:new_seed).and_return(1, 2)

  expect(described_class.generate).not_to eq(described_class.generate)
end
```

I'm tempted to leave it as-is. What do you think?

Reply:
I think what I was imagining was that if you could set the seed in test, then you'd always get the same sequence of generated numbers: 

```
irb(main):001> seed = 123
=> 123
irb(main):002> rng = Random.new(seed)
=> #<Random:0x0000000110093088>
irb(main):003> rng.rand(100)
=> 66
irb(main):004> rng.rand(100)
=> 92
```

so you could write the test to expect those specific numbers (or expect that they're different to one another as you do now). 

But the chance of generating the same number twice in a row is 1:1000000 * 1:1000000 or 1 in a billion! So I don't think we do really need to worry about it 😂

Reply:
Ah, I see. So use the same seed to guarantee a different number each time, rather than two different seeds.

However, I'm going to leave it as-is given that the chance of collision seems so small. Thank you!